python/ray/data/_internal/plan.py

```diff
@@ -328,33 +328,62 @@ def execute(
             The blocks of the output dataset.
         """
         if not self.has_computed_output():
-            blocks, stats, stages = self._optimize()
             context = DatasetContext.get_current()
-            for stage_idx, stage in enumerate(stages):
-                if allow_clear_input_blocks:
-                    clear_input_blocks = self._should_clear_input_blocks(
-                        blocks, stage_idx
-                    )
-                else:
-                    clear_input_blocks = False
-                stats_builder = stats.child_builder(stage.name)
-                blocks, stage_info = stage(
-                    blocks, clear_input_blocks, self._run_by_consumer
+
+            # Read stage is handled with the legacy execution impl for now.
+            if (
+                context.new_execution_backend
+                and not self.is_read_stage_equivalent()
+                and self._stages_after_snapshot
+            ):
+                from ray.data._internal.execution.bulk_executor import BulkExecutor
+                from ray.data._internal.execution.interfaces import ExecutionOptions
+                from ray.data._internal.execution.legacy_compat import (
+                    execute_to_legacy_block_list,
                 )
-                if stage_info:
-                    stats = stats_builder.build_multistage(stage_info)
-                else:
-                    stats = stats_builder.build(blocks)
-                stats.dataset_uuid = uuid.uuid4().hex
 
-                stats_summary_string = stats.summary_string(include_parent=False)
-                logger.get_logger(log_to_stdout=context.enable_auto_log_stats).info(
-                    stats_summary_string,
+                executor = BulkExecutor(ExecutionOptions())
+                blocks = execute_to_legacy_block_list(
+                    executor,
+                    self,
+                    allow_clear_input_blocks=allow_clear_input_blocks,
+                    dataset_uuid=self._dataset_uuid,
                 )
+                # TODO(ekl) we shouldn't need to set this in the future once we move
```

Comment:
can we file a github issue for this TODO?

Reply:
I'll do this after the REP is approved.